# Data Cleaning Challenge

#### Import necessary libraries

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import datetime 
import matplotlib.pyplot as plt

#### Import dataset to Jupyter Notebook

In [2]:
file_location = 'C:\\Users\\ragir\\OneDrive\\Desktop\\DATA ANALYTICS\\Portfolio Projects\\Data Cleaning Challenge\\'

fifa_data = pd.read_csv(file_location + 'fifa21 raw data v2.csv', low_memory=False)


In [3]:
# Let us view the dataset
fifa_data.head()

ID               Name                      LongName  \
0  158023           L. Messi                  Lionel Messi   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   
2  200389           J. Oblak                     Jan Oblak   
3  192985       K. De Bruyne               Kevin De Bruyne   
4  190871          Neymar Jr    Neymar da Silva Santos Jr.   

                                           photoUrl  \
0  https://cdn.sofifa.com/players/158/023/21_60.png   
1  https://cdn.sofifa.com/players/020/801/21_60.png   
2  https://cdn.sofifa.com/players/200/389/21_60.png   
3  https://cdn.sofifa.com/players/192/985/21_60.png   
4  https://cdn.sofifa.com/players/190/871/21_60.png   

                                           playerUrl Nationality  Age  ↓OVA  \
0  http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33    93   
1  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35    92   
2  http://sofifa.com/player/200389/jan-oblak/210006/    Slovenia   27    91   
3  http://sofifa.com/player/192985/kevin-de-bruyn...     Belgium   29    91   
4  http://sofifa.com/player/190871/neymar-da-silv...      Brazil   28    91   

   POT                         Club  ...     A/W     D/W   IR PAC SHO  PAS  \
0   93         \n\n\n\nFC Barcelona  ...  Medium     Low  5 ★  85  92   91   
1   92             \n\n\n\nJuventus  ...    High     Low  5 ★  89  93   81   
2   93      \n\n\n\nAtlético Madrid  ...  Medium  Medium  3 ★  87  92   78   
3   91      \n\n\n\nManchester City  ...    High    High  4 ★  76  86   93   
4   91  \n\n\n\nParis Saint-Germain  ...    High  Medium  5 ★  91  85   86   

  DRI DEF PHY Hits  
0  95  38  65  771  
1  89  35  77  562  
2  90  52  90  150  
3  88  64  78  207  
4  94  36  59  595  

[5 rows x 77 columns]

In [4]:
fifa_data.tail()

ID          Name        LongName  \
18974  247223        Xia Ao          Ao Xia   
18975  258760      B. Hough       Ben Hough   
18976  252757   R. McKinley  Ronan McKinley   
18977  243790  Wang Zhen'ao    Zhen'ao Wang   
18978  252520     Zhou Xiao       Xiao Zhou   

                                               photoUrl  \
18974  https://cdn.sofifa.com/players/247/223/21_60.png   
18975  https://cdn.sofifa.com/players/258/760/21_60.png   
18976  https://cdn.sofifa.com/players/252/757/21_60.png   
18977  https://cdn.sofifa.com/players/243/790/21_60.png   
18978  https://cdn.sofifa.com/players/252/520/21_60.png   

                                               playerUrl Nationality  Age  \
18974     http://sofifa.com/player/247223/ao-xia/210006/    China PR   21   
18975  http://sofifa.com/player/258760/ben-hough/210006/     England   17   
18976  http://sofifa.com/player/252757/ronan-mckinley...     England   18   
18977  http://sofifa.com/player/243790/zhenao-wang/21...    China PR   20   
18978  http://sofifa.com/player/252520/xiao-zhou/210006/    China PR   21   

       ↓OVA  POT                      Club  ...     A/W     D/W   IR PAC SHO  \
18974    47   55        \n\n\n\nWuhan Zall  ...  Medium  Medium  1 ★  64  28   
18975    47   67   \n\n\n\nOldham Athletic  ...  Medium  Medium  1 ★  64  40   
18976    47   65        \n\n\n\nDerry City  ...  Medium  Medium  1 ★  63  39   
18977    47   57  \n\n\n\nDalian YiFang FC  ...  Medium  Medium  1 ★  58  49   
18978    47   57  \n\n\n\nDalian YiFang FC  ...  Medium  Medium  1 ★  62  22   

       PAS DRI DEF PHY Hits  
18974   26  38  48  51  NaN  
18975   48  49  35  45  NaN  
18976   44  46  40  53  NaN  
18977   41  49  30  44  NaN  
18978   39  42  45  55  NaN  

[5 rows x 77 columns]

In [5]:
# View number of rows and coumns in the dataset
# 18979 rows and 77 columns

fifa_data.shape

(18979, 77)

In [6]:
# Check and delete duplicates in the dataset

# Check for duplicates
print(fifa_data.duplicated().sum())

# Drop duplicates
fifa_data.drop_duplicates(inplace=True)

# Check for duplicates after dropping
print(fifa_data.duplicated().sum())

0
0


There are no duplicates in the dataset

In [7]:
# Let's view all the column titles
fifa_data.columns

Index(['ID', 'Name', 'LongName', 'photoUrl', 'playerUrl', 'Nationality', 'Age',
       '↓OVA', 'POT', 'Club', 'Contract', 'Positions', 'Height', 'Weight',
       'Preferred Foot', 'BOV', 'Best Position', 'Joined', 'Loan Date End',
       'Value', 'Wage', 'Release Clause', 'Attacking', 'Crossing', 'Finishing',
       'Heading Accuracy', 'Short Passing', 'Volleys', 'Skill', 'Dribbling',
       'Curve', 'FK Accuracy', 'Long Passing', 'Ball Control', 'Movement',
       'Acceleration', 'Sprint Speed', 'Agility', 'Reactions', 'Balance',
       'Power', 'Shot Power', 'Jumping', 'Stamina', 'Strength', 'Long Shots',
       'Mentality', 'Aggression', 'Interceptions', 'Positioning', 'Vision',
       'Penalties', 'Composure', 'Defending', 'Marking', 'Standing Tackle',
       'Sliding Tackle', 'Goalkeeping', 'GK Diving', 'GK Handling',
       'GK Kicking', 'GK Positioning', 'GK Reflexes', 'Total Stats',
       'Base Stats', 'W/F', 'SM', 'A/W', 'D/W', 'IR', 'PAC', 'SHO', 'PAS',
       'DRI', 'DEF', 

In [8]:
# Rename the ↓OVA column to OVA

fifa_data = fifa_data.rename(columns={'↓OVA': 'OVA'})

In [9]:
# Let's clean the ID column

# Check for datatype
print(fifa_data['ID'].dtype)

# Check for any missing values
print(fifa_data['ID'].isnull().sum())


int64
0


The ID column has int values and no mssing values

In [10]:
# Let's clean the Name column

# Check for datatype
print(fifa_data['Name'].dtype)

# Check for any missing values
print(fifa_data['Name'].isnull().sum())

# Check for numerical data in the Name column
numerical_data = fifa_data['Name'].str.contains('\d+')

# Print rows with numerical data in the Name column
print(fifa_data[numerical_data])



object
0
Empty DataFrame
Columns: [ID, Name, LongName, photoUrl, playerUrl, Nationality, Age, OVA, POT, Club, Contract, Positions, Height, Weight, Preferred Foot, BOV, Best Position, Joined, Loan Date End, Value, Wage, Release Clause, Attacking, Crossing, Finishing, Heading Accuracy, Short Passing, Volleys, Skill, Dribbling, Curve, FK Accuracy, Long Passing, Ball Control, Movement, Acceleration, Sprint Speed, Agility, Reactions, Balance, Power, Shot Power, Jumping, Stamina, Strength, Long Shots, Mentality, Aggression, Interceptions, Positioning, Vision, Penalties, Composure, Defending, Marking, Standing Tackle, Sliding Tackle, Goalkeeping, GK Diving, GK Handling, GK Kicking, GK Positioning, GK Reflexes, Total Stats, Base Stats, W/F, SM, A/W, D/W, IR, PAC, SHO, PAS, DRI, DEF, PHY, Hits]
Index: []

[0 rows x 77 columns]


The Name column is of object datatype and has 0 missing values and no number values in the column

In [11]:
# Let's clean the LongName column

# Check for datatype
print(fifa_data['LongName'].dtype)

# Check for any missing values
print(fifa_data['LongName'].isnull().sum())

# Check for numerical data in the Name column
numerical_data = fifa_data['LongName'].str.contains('\d+')

# Print rows with numerical data in the Name column
print(fifa_data[numerical_data])

object
0
Empty DataFrame
Columns: [ID, Name, LongName, photoUrl, playerUrl, Nationality, Age, OVA, POT, Club, Contract, Positions, Height, Weight, Preferred Foot, BOV, Best Position, Joined, Loan Date End, Value, Wage, Release Clause, Attacking, Crossing, Finishing, Heading Accuracy, Short Passing, Volleys, Skill, Dribbling, Curve, FK Accuracy, Long Passing, Ball Control, Movement, Acceleration, Sprint Speed, Agility, Reactions, Balance, Power, Shot Power, Jumping, Stamina, Strength, Long Shots, Mentality, Aggression, Interceptions, Positioning, Vision, Penalties, Composure, Defending, Marking, Standing Tackle, Sliding Tackle, Goalkeeping, GK Diving, GK Handling, GK Kicking, GK Positioning, GK Reflexes, Total Stats, Base Stats, W/F, SM, A/W, D/W, IR, PAC, SHO, PAS, DRI, DEF, PHY, Hits]
Index: []

[0 rows x 77 columns]


In [12]:
# Let's clean the photoUrl Column

# Check for datatype
print(fifa_data['photoUrl'].dtype)

# Check for any missing values
print(fifa_data['photoUrl'].isnull().sum())

# Check photo extension of each photo
fifa_data.photoUrl.head(10)


object
0


0    https://cdn.sofifa.com/players/158/023/21_60.png
1    https://cdn.sofifa.com/players/020/801/21_60.png
2    https://cdn.sofifa.com/players/200/389/21_60.png
3    https://cdn.sofifa.com/players/192/985/21_60.png
4    https://cdn.sofifa.com/players/190/871/21_60.png
5    https://cdn.sofifa.com/players/188/545/21_60.png
6    https://cdn.sofifa.com/players/209/331/21_60.png
7    https://cdn.sofifa.com/players/212/831/21_60.png
8    https://cdn.sofifa.com/players/231/747/21_60.png
9    https://cdn.sofifa.com/players/192/448/21_60.png
Name: photoUrl, dtype: object

In [13]:
# Check if all the photos are .png files
is_png = fifa_data['photoUrl'].str.endswith('.png').all()

# Print result
print(is_png)

True


We use the str.endswith() method to check if all photo URLs in the photoUrl column end with the .png extension. We pass the string .png as the suffix to search for. The str.endswith() method returns a boolean series indicating whether each value in the photoUrl column ends with the suffix or not. We use the all() function to check if all the values are True.

All the photos in the photoUrl column are .png files

In [14]:
# Let's clean the playerUrl column

# Check for datatype
print(fifa_data['playerUrl'].dtype)

# Check for any missing values
print(fifa_data['playerUrl'].isnull().sum())

# Check playerUrl data
fifa_data.playerUrl.head(10)


object
0


0    http://sofifa.com/player/158023/lionel-messi/2...
1    http://sofifa.com/player/20801/c-ronaldo-dos-s...
2    http://sofifa.com/player/200389/jan-oblak/210006/
3    http://sofifa.com/player/192985/kevin-de-bruyn...
4    http://sofifa.com/player/190871/neymar-da-silv...
5    http://sofifa.com/player/188545/robert-lewando...
6    http://sofifa.com/player/209331/mohamed-salah/...
7    http://sofifa.com/player/212831/alisson-ramses...
8    http://sofifa.com/player/231747/kylian-mbappe/...
9    http://sofifa.com/player/192448/marc-andre-ter...
Name: playerUrl, dtype: object

There is no missing value in the playerUrl column

In [15]:
# Let's clean the Nationality column

# Check for datatype
print(fifa_data['Nationality'].dtype)

# Check for any missing values
print(fifa_data['Nationality'].isnull().sum())

# Count unique values in the Nationality column
num_unique = fifa_data['Nationality'].nunique()

# Print result
print(num_unique)

# Check Nationality data
fifa_data.Nationality.unique()

object
0
164


array(['Argentina', 'Portugal', 'Slovenia', 'Belgium', 'Brazil', 'Poland',
       'Egypt', 'France', 'Germany', 'Netherlands', 'Senegal', 'Spain',
       'England', 'Scotland', 'Korea Republic', 'Costa Rica', 'Italy',
       'Gabon', 'Croatia', 'Uruguay', 'Switzerland', 'Serbia', 'Slovakia',
       'Morocco', 'Algeria', 'Denmark', 'Hungary', 'Bosnia Herzegovina',
       'Nigeria', 'Cameroon', 'Norway', 'Ghana', 'Mexico', 'Austria',
       'Albania', 'Colombia', 'Chile', 'Ivory Coast', 'Greece', 'Finland',
       'Wales', 'Sweden', 'Togo', 'Czech Republic', 'Russia', 'Venezuela',
       'Canada', 'United States', 'Guinea', 'Montenegro', 'Israel',
       'Republic of Ireland', 'Ukraine', 'Turkey', 'Ecuador', 'Jamaica',
       'DR Congo', 'Australia', 'China PR', 'Armenia', 'Northern Ireland',
       'North Macedonia', 'Kosovo', 'Mali', 'Peru',
       'Central African Republic', 'Iceland', 'Burkina Faso', 'Paraguay',
       'Japan', 'Romania', 'New Zealand', 'Iran', 'Angola', 'Tunisia',
 

There are 164 different Nationalities in the Natioanlity column of the dataset

In [16]:
# Create a new dataframe to add the cleaned columns above: ID, Name, LongName, photoURl, playerUrl, Nationality

# Clean columns
clean_data = fifa_data[['ID', 'Name', 'LongName', 'photoUrl', 'playerUrl', 'Nationality']]

# Create new dataframe
fifa_clean = pd.DataFrame(clean_data)

# Print new dataframe
print(fifa_clean)


           ID               Name                      LongName  \
0      158023           L. Messi                  Lionel Messi   
1       20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   
2      200389           J. Oblak                     Jan Oblak   
3      192985       K. De Bruyne               Kevin De Bruyne   
4      190871          Neymar Jr    Neymar da Silva Santos Jr.   
...       ...                ...                           ...   
18974  247223             Xia Ao                        Ao Xia   
18975  258760           B. Hough                     Ben Hough   
18976  252757        R. McKinley                Ronan McKinley   
18977  243790       Wang Zhen'ao                  Zhen'ao Wang   
18978  252520          Zhou Xiao                     Xiao Zhou   

                                               photoUrl  \
0      https://cdn.sofifa.com/players/158/023/21_60.png   
1      https://cdn.sofifa.com/players/020/801/21_60.png   
2      https://cdn.sofifa.com/

In [17]:
# Let us separate the clean columns from the dirty ones by creating a new dataset

other_data = fifa_data.drop(['ID', 'Name', 'LongName', 'photoUrl', 'playerUrl', 'Nationality'], axis=1)
fifa_dirty = pd.DataFrame(other_data)
print(fifa_dirty)

       Age  OVA  POT                         Club     Contract   Positions  \
0       33   93   93         \n\n\n\nFC Barcelona  2004 ~ 2021  RW, ST, CF   
1       35   92   92             \n\n\n\nJuventus  2018 ~ 2022      ST, LW   
2       27   91   93      \n\n\n\nAtlético Madrid  2014 ~ 2023          GK   
3       29   91   91      \n\n\n\nManchester City  2015 ~ 2023     CAM, CM   
4       28   91   91  \n\n\n\nParis Saint-Germain  2017 ~ 2022     LW, CAM   
...    ...  ...  ...                          ...          ...         ...   
18974   21   47   55           \n\n\n\nWuhan Zall  2018 ~ 2022          CB   
18975   17   47   67      \n\n\n\nOldham Athletic  2020 ~ 2021          CM   
18976   18   47   65           \n\n\n\nDerry City  2019 ~ 2020          CM   
18977   20   47   57     \n\n\n\nDalian YiFang FC  2020 ~ 2022          RW   
18978   21   47   57     \n\n\n\nDalian YiFang FC  2019 ~ 2023      CB, LB   

      Height Weight Preferred Foot  BOV  ...     A/W     D/W   

In [18]:
# Let us separate and group the remaining columns into their datatypes for easy cleaning

# Create empty lists for each datatype
int_cols = []
str_cols = []
other_cols = []

# Loop through columns and categorize by datatype
for col in fifa_dirty.columns:
    if fifa_dirty[col].dtype == 'int64':
        int_cols.append(col)
    elif fifa_dirty[col].dtype == 'object':
        str_cols.append(col)
    else:
        other_cols.append(col)

In [19]:
# Lets look at the integer columns

print(int_cols)

print("Count: ", len(int_cols))

# 53 columns are integers. These are easier to clean since they have less issues

['Age', 'OVA', 'POT', 'BOV', 'Attacking', 'Crossing', 'Finishing', 'Heading Accuracy', 'Short Passing', 'Volleys', 'Skill', 'Dribbling', 'Curve', 'FK Accuracy', 'Long Passing', 'Ball Control', 'Movement', 'Acceleration', 'Sprint Speed', 'Agility', 'Reactions', 'Balance', 'Power', 'Shot Power', 'Jumping', 'Stamina', 'Strength', 'Long Shots', 'Mentality', 'Aggression', 'Interceptions', 'Positioning', 'Vision', 'Penalties', 'Composure', 'Defending', 'Marking', 'Standing Tackle', 'Sliding Tackle', 'Goalkeeping', 'GK Diving', 'GK Handling', 'GK Kicking', 'GK Positioning', 'GK Reflexes', 'Total Stats', 'Base Stats', 'PAC', 'SHO', 'PAS', 'DRI', 'DEF', 'PHY']
Count:  53


In [20]:
# Let's look at the int_cols as a group


fifa_dirty[int_cols].isna().any()
fifa_dirty[int_cols].isnull().any()

# we don't have any missing value
# lets look at the data types

fifa_dirty[int_cols].dtypes

Age                 int64
OVA                 int64
POT                 int64
BOV                 int64
Attacking           int64
Crossing            int64
Finishing           int64
Heading Accuracy    int64
Short Passing       int64
Volleys             int64
Skill               int64
Dribbling           int64
Curve               int64
FK Accuracy         int64
Long Passing        int64
Ball Control        int64
Movement            int64
Acceleration        int64
Sprint Speed        int64
Agility             int64
Reactions           int64
Balance             int64
Power               int64
Shot Power          int64
Jumping             int64
Stamina             int64
Strength            int64
Long Shots          int64
Mentality           int64
Aggression          int64
Interceptions       int64
Positioning         int64
Vision              int64
Penalties           int64
Composure           int64
Defending           int64
Marking             int64
Standing Tackle     int64
Sliding Tack

In [21]:
# Describing our data is the best way to check out for any issues in the int_cols

fifa_dirty[int_cols].describe()



Age           OVA           POT           BOV     Attacking  \
count  18979.000000  18979.000000  18979.000000  18979.000000  18979.000000   
mean      25.194109     65.718636     71.136414     66.751726    248.938142   
std        4.710520      6.968999      6.114635      6.747193     74.299428   
min       16.000000     47.000000     47.000000     48.000000     42.000000   
25%       21.000000     61.000000     67.000000     62.000000    222.000000   
50%       25.000000     66.000000     71.000000     67.000000    263.000000   
75%       29.000000     70.000000     75.000000     71.000000    297.000000   
max       53.000000     93.000000     95.000000     93.000000    437.000000   

           Crossing     Finishing  Heading Accuracy  Short Passing  \
count  18979.000000  18979.000000      18979.000000   18979.000000   
mean      49.688392     45.842405         51.942726      58.768112   
std       18.131153     19.567081         17.294409      14.519106   
min        6.000000      3.000000          5.000000       7.000000   
25%       38.000000     30.000000         44.000000      54.000000   
50%       54.000000     49.000000         55.000000      62.000000   
75%       63.000000     62.000000         64.000000      68.000000   
max       94.000000     95.000000         93.000000      94.000000   

            Volleys  ...  GK Positioning   GK Reflexes   Total Stats  \
count  18979.000000  ...    18979.000000  18979.000000  18979.000000   
mean      42.696507  ...       16.217187     16.519627   1595.286949   
std       17.646937  ...       17.002239     17.854079    269.874789   
min        3.000000  ...        2.000000      2.000000    747.000000   
25%       30.000000  ...        8.000000      8.000000   1452.000000   
50%       44.000000  ...       11.000000     11.000000   1627.000000   
75%       56.000000  ...       14.000000     14.000000   1781.000000   
max       90.000000  ...       91.000000     90.000000   2316.000000   

         Base Stats           PAC           SHO           PAS           DRI  \
count  18979.000000  18979.000000  18979.000000  18979.000000  18979.000000   
mean     355.702197     67.453975     53.457031     57.681016     62.875020   
std       40.761117     10.677859     13.827425     10.081857      9.927415   
min      232.000000     25.000000     16.000000     25.000000     25.000000   
25%      327.000000     61.000000     44.000000     51.000000     57.000000   
50%      356.000000     68.000000     56.000000     58.000000     64.000000   
75%      384.000000     75.000000     64.000000     64.000000     69.000000   
max      498.000000     96.000000     93.000000     93.000000     95.000000   

                DEF           PHY  
count  18979.000000  18979.000000  
mean      49.866221     64.368934  
std       16.443213      9.601883  
min       12.000000     28.000000  
25%       35.000000     58.000000  
50%       53.000000     65.000000  
75%       63.000000     71.000000  
max       91.000000     91.000000  

[8 rows x 53 columns]

From the result we have no negative numbers, hence there are no issues with the int_cols

In [22]:
# Let's add the int_cols into fifa_clean dataframe

# Create a list of integer columns in fifa_dirty
int_cols = [col for col in fifa_dirty.columns if fifa_dirty[col].dtype == 'int64']

# Select the integer columns from fifa_dirty
fifa_dirty_int = fifa_dirty[int_cols]

# Concatenate the integer columns with fifa_clean
fifa_clean = pd.concat([fifa_clean, fifa_dirty_int], axis=1)

# Verify that the integer columns have been added to fifa_clean
print(fifa_clean.info())


<class 'pandas.core.frame.DataFrame'>
Int64Index: 18979 entries, 0 to 18978
Data columns (total 59 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   ID                18979 non-null  int64 
 1   Name              18979 non-null  object
 2   LongName          18979 non-null  object
 3   photoUrl          18979 non-null  object
 4   playerUrl         18979 non-null  object
 5   Nationality       18979 non-null  object
 6   Age               18979 non-null  int64 
 7   OVA               18979 non-null  int64 
 8   POT               18979 non-null  int64 
 9   BOV               18979 non-null  int64 
 10  Attacking         18979 non-null  int64 
 11  Crossing          18979 non-null  int64 
 12  Finishing         18979 non-null  int64 
 13  Heading Accuracy  18979 non-null  int64 
 14  Short Passing     18979 non-null  int64 
 15  Volleys           18979 non-null  int64 
 16  Skill             18979 non-null  int64 
 17  Dribbling   

In [23]:
# Let's look at the stri_cols and other_cols

print(str_cols)

print("Count: ", len(str_cols))

print(other_cols)

print("Count: ", len(other_cols))

# We have 18 str_cols and 0 other_cols in the fifa_dirty dataframe 

['Club', 'Contract', 'Positions', 'Height', 'Weight', 'Preferred Foot', 'Best Position', 'Joined', 'Loan Date End', 'Value', 'Wage', 'Release Clause', 'W/F', 'SM', 'A/W', 'D/W', 'IR', 'Hits']
Count:  18
[]
Count:  0


In [24]:
# Let's clean the Club column

fifa_dirty['Club']


0               \n\n\n\nFC Barcelona
1                   \n\n\n\nJuventus
2            \n\n\n\nAtlético Madrid
3            \n\n\n\nManchester City
4        \n\n\n\nParis Saint-Germain
                    ...             
18974             \n\n\n\nWuhan Zall
18975        \n\n\n\nOldham Athletic
18976             \n\n\n\nDerry City
18977       \n\n\n\nDalian YiFang FC
18978       \n\n\n\nDalian YiFang FC
Name: Club, Length: 18979, dtype: object

In [25]:
# Clean the Club column by removing the whitespaces and other characters

fifa_dirty['Club'] = fifa_dirty['Club'].str.strip()

fifa_dirty['Club']


0               FC Barcelona
1                   Juventus
2            Atlético Madrid
3            Manchester City
4        Paris Saint-Germain
                ...         
18974             Wuhan Zall
18975        Oldham Athletic
18976             Derry City
18977       Dalian YiFang FC
18978       Dalian YiFang FC
Name: Club, Length: 18979, dtype: object

In [26]:
# Add the Club column to the fifa_clean dataframe

fifa_clean = fifa_clean.join(fifa_dirty['Club'])


In [27]:
# Verify that the Club column has been added to fifa_clean

print(fifa_clean.info())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 18979 entries, 0 to 18978
Data columns (total 60 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   ID                18979 non-null  int64 
 1   Name              18979 non-null  object
 2   LongName          18979 non-null  object
 3   photoUrl          18979 non-null  object
 4   playerUrl         18979 non-null  object
 5   Nationality       18979 non-null  object
 6   Age               18979 non-null  int64 
 7   OVA               18979 non-null  int64 
 8   POT               18979 non-null  int64 
 9   BOV               18979 non-null  int64 
 10  Attacking         18979 non-null  int64 
 11  Crossing          18979 non-null  int64 
 12  Finishing         18979 non-null  int64 
 13  Heading Accuracy  18979 non-null  int64 
 14  Short Passing     18979 non-null  int64 
 15  Volleys           18979 non-null  int64 
 16  Skill             18979 non-null  int64 
 17  Dribbling   

In [28]:
# Let's clean the Positions column

fifa_dirty['Positions'] = fifa_dirty['Positions'].str.replace(', ', '/')


In [29]:
# Add the Positions column to the fifa_clean dataframe

fifa_clean = fifa_clean.join(fifa_dirty['Positions'])

# Verify that the Positions column has been added to fifa_clean

print(fifa_clean.info())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 18979 entries, 0 to 18978
Data columns (total 61 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   ID                18979 non-null  int64 
 1   Name              18979 non-null  object
 2   LongName          18979 non-null  object
 3   photoUrl          18979 non-null  object
 4   playerUrl         18979 non-null  object
 5   Nationality       18979 non-null  object
 6   Age               18979 non-null  int64 
 7   OVA               18979 non-null  int64 
 8   POT               18979 non-null  int64 
 9   BOV               18979 non-null  int64 
 10  Attacking         18979 non-null  int64 
 11  Crossing          18979 non-null  int64 
 12  Finishing         18979 non-null  int64 
 13  Heading Accuracy  18979 non-null  int64 
 14  Short Passing     18979 non-null  int64 
 15  Volleys           18979 non-null  int64 
 16  Skill             18979 non-null  int64 
 17  Dribbling   

In [30]:
print(fifa_clean['Positions'])


0        RW/ST/CF
1           ST/LW
2              GK
3          CAM/CM
4          LW/CAM
           ...   
18974          CB
18975          CM
18976          CM
18977          RW
18978       CB/LB
Name: Positions, Length: 18979, dtype: object


In [31]:
# Let's clean the Height column

# Count unique values in the Height column
num_unique = fifa_dirty['Height'].nunique()

# Print result
print(num_unique)

# Check Height data
fifa_dirty.Height.unique()

62


array(['170cm', '187cm', '188cm', '181cm', '175cm', '184cm', '191cm',
       '178cm', '193cm', '185cm', '199cm', '173cm', '168cm', '176cm',
       '177cm', '183cm', '180cm', '189cm', '179cm', '195cm', '172cm',
       '182cm', '186cm', '192cm', '165cm', '194cm', '167cm', '196cm',
       '163cm', '190cm', '174cm', '169cm', '171cm', '197cm', '200cm',
       '166cm', '6\'2"', '164cm', '198cm', '6\'3"', '6\'5"', '5\'11"',
       '6\'4"', '6\'1"', '6\'0"', '5\'10"', '5\'9"', '5\'6"', '5\'7"',
       '5\'4"', '201cm', '158cm', '162cm', '161cm', '160cm', '203cm',
       '157cm', '156cm', '202cm', '159cm', '206cm', '155cm'], dtype=object)

We see that some data is in centimeters while there's other data in feet.

In [32]:
# Let's replace all data in feet into centimeters- replace ' with . and " with ''

fifa_dirty['Height'][fifa_dirty.Height.str.contains("'")]

fifa_dirty['Height'] = fifa_dirty.Height.str.replace("'", '.')
fifa_dirty['Height'] = fifa_dirty.Height.str.replace('"', '')

fifa_dirty.Height.unique()

array(['170cm', '187cm', '188cm', '181cm', '175cm', '184cm', '191cm',
       '178cm', '193cm', '185cm', '199cm', '173cm', '168cm', '176cm',
       '177cm', '183cm', '180cm', '189cm', '179cm', '195cm', '172cm',
       '182cm', '186cm', '192cm', '165cm', '194cm', '167cm', '196cm',
       '163cm', '190cm', '174cm', '169cm', '171cm', '197cm', '200cm',
       '166cm', '6.2', '164cm', '198cm', '6.3', '6.5', '5.11', '6.4',
       '6.1', '6.0', '5.10', '5.9', '5.6', '5.7', '5.4', '201cm', '158cm',
       '162cm', '161cm', '160cm', '203cm', '157cm', '156cm', '202cm',
       '159cm', '206cm', '155cm'], dtype=object)

In [33]:
# Let's convert the data in feet into centimeters for consistency

def convert_height(height):
    if 'cm' in height:
        # height is already in centimeters
        return int(height.replace('cm', ''))
    else:
        # split the height in feet and inches using a period
        feet, inches = height.split(".")
        # convert feet and inches to centimeters
        height_cm = int(feet) * 30.48 + int(inches) * 2.54
        return height_cm

fifa_dirty['Height'] = fifa_dirty['Height'].apply(convert_height)


In [34]:
fifa_dirty.Height.unique()

array([170.  , 187.  , 188.  , 181.  , 175.  , 184.  , 191.  , 178.  ,
       193.  , 185.  , 199.  , 173.  , 168.  , 176.  , 177.  , 183.  ,
       180.  , 189.  , 179.  , 195.  , 172.  , 182.  , 186.  , 192.  ,
       165.  , 194.  , 167.  , 196.  , 163.  , 190.  , 174.  , 169.  ,
       171.  , 197.  , 200.  , 166.  , 187.96, 164.  , 198.  , 190.5 ,
       195.58, 180.34, 193.04, 185.42, 182.88, 177.8 , 175.26, 167.64,
       170.18, 162.56, 201.  , 158.  , 162.  , 161.  , 160.  , 203.  ,
       157.  , 156.  , 202.  , 159.  , 206.  , 155.  ])

In [35]:
# Add cm suffix to the Height column and in zero decimal places

fifa_dirty['Height'] = fifa_dirty['Height'].round(0).apply(lambda x: f"{int(x)}cm")


In [36]:
fifa_dirty.Height.unique()

array(['170cm', '187cm', '188cm', '181cm', '175cm', '184cm', '191cm',
       '178cm', '193cm', '185cm', '199cm', '173cm', '168cm', '176cm',
       '177cm', '183cm', '180cm', '189cm', '179cm', '195cm', '172cm',
       '182cm', '186cm', '192cm', '165cm', '194cm', '167cm', '196cm',
       '163cm', '190cm', '174cm', '169cm', '171cm', '197cm', '200cm',
       '166cm', '164cm', '198cm', '201cm', '158cm', '162cm', '161cm',
       '160cm', '203cm', '157cm', '156cm', '202cm', '159cm', '206cm',
       '155cm'], dtype=object)

In [37]:
# Add Height column to fifa_clean dataframe

fifa_clean = fifa_clean.join(fifa_dirty['Height'])

# Verify that the Height column has been added to fifa_clean

print(fifa_clean.info())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 18979 entries, 0 to 18978
Data columns (total 62 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   ID                18979 non-null  int64 
 1   Name              18979 non-null  object
 2   LongName          18979 non-null  object
 3   photoUrl          18979 non-null  object
 4   playerUrl         18979 non-null  object
 5   Nationality       18979 non-null  object
 6   Age               18979 non-null  int64 
 7   OVA               18979 non-null  int64 
 8   POT               18979 non-null  int64 
 9   BOV               18979 non-null  int64 
 10  Attacking         18979 non-null  int64 
 11  Crossing          18979 non-null  int64 
 12  Finishing         18979 non-null  int64 
 13  Heading Accuracy  18979 non-null  int64 
 14  Short Passing     18979 non-null  int64 
 15  Volleys           18979 non-null  int64 
 16  Skill             18979 non-null  int64 
 17  Dribbling   

In [38]:
# Let's clean the Weight column

# Count unique values in the Height column
num_unique = fifa_dirty['Weight'].nunique()

# Print result
print(num_unique)

# Check Height data
fifa_dirty.Weight.unique()

79


array(['72kg', '83kg', '87kg', '70kg', '68kg', '80kg', '71kg', '91kg',
       '73kg', '85kg', '92kg', '69kg', '84kg', '96kg', '81kg', '82kg',
       '75kg', '86kg', '89kg', '74kg', '76kg', '64kg', '78kg', '90kg',
       '66kg', '60kg', '94kg', '79kg', '67kg', '65kg', '59kg', '61kg',
       '93kg', '88kg', '97kg', '77kg', '62kg', '63kg', '95kg', '100kg',
       '58kg', '183lbs', '179lbs', '172lbs', '196lbs', '176lbs', '185lbs',
       '170lbs', '203lbs', '168lbs', '161lbs', '146lbs', '130lbs',
       '190lbs', '174lbs', '148lbs', '165lbs', '159lbs', '192lbs',
       '181lbs', '139lbs', '154lbs', '157lbs', '163lbs', '98kg', '103kg',
       '99kg', '102kg', '56kg', '101kg', '57kg', '55kg', '104kg', '107kg',
       '110kg', '53kg', '50kg', '54kg', '52kg'], dtype=object)

In [39]:
# Some data is in kgs and other in lbs.  Let's convert the data in lbs into kgs for consistency
# Loop through the "Weight" column and convert the values in lbs to kgs
for i in range(len(fifa_dirty)):
    if fifa_dirty.loc[i, 'Weight'][-3:] == 'lbs':
        weight_lb = float(fifa_dirty.loc[i, 'Weight'][:-3])
        weight_kg = weight_lb * 0.45359237
        fifa_dirty.loc[i, 'Weight'] = str(round(weight_kg)) + 'kg'


# Check Weight data
fifa_dirty.Weight.unique()

array(['72kg', '83kg', '87kg', '70kg', '68kg', '80kg', '71kg', '91kg',
       '73kg', '85kg', '92kg', '69kg', '84kg', '96kg', '81kg', '82kg',
       '75kg', '86kg', '89kg', '74kg', '76kg', '64kg', '78kg', '90kg',
       '66kg', '60kg', '94kg', '79kg', '67kg', '65kg', '59kg', '61kg',
       '93kg', '88kg', '97kg', '77kg', '62kg', '63kg', '95kg', '100kg',
       '58kg', '98kg', '103kg', '99kg', '102kg', '56kg', '101kg', '57kg',
       '55kg', '104kg', '107kg', '110kg', '53kg', '50kg', '54kg', '52kg'],
      dtype=object)

In [40]:
# Add Weight column to fifa_clean dataframe

fifa_clean = fifa_clean.join(fifa_dirty['Weight'])

# Verify that the Weight column has been added to fifa_clean

print(fifa_clean.info())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 18979 entries, 0 to 18978
Data columns (total 63 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   ID                18979 non-null  int64 
 1   Name              18979 non-null  object
 2   LongName          18979 non-null  object
 3   photoUrl          18979 non-null  object
 4   playerUrl         18979 non-null  object
 5   Nationality       18979 non-null  object
 6   Age               18979 non-null  int64 
 7   OVA               18979 non-null  int64 
 8   POT               18979 non-null  int64 
 9   BOV               18979 non-null  int64 
 10  Attacking         18979 non-null  int64 
 11  Crossing          18979 non-null  int64 
 12  Finishing         18979 non-null  int64 
 13  Heading Accuracy  18979 non-null  int64 
 14  Short Passing     18979 non-null  int64 
 15  Volleys           18979 non-null  int64 
 16  Skill             18979 non-null  int64 
 17  Dribbling   

In [41]:
# Let's check for nulls and NA values in the Height and Weight column in fifa_clean dataframe

# Check for nulls or NA values in the "Height" column

if fifa_clean['Height'].isnull().sum() > 0 or fifa_clean['Height'].isna().sum() > 0:
    print("There are nulls or NA values in the 'Height' column.")
else:
    print("There are no nulls or NA values in the 'Height' column.")

# Check for nulls or NA values in the "Weight" column
if fifa_clean['Weight'].isnull().sum() > 0 or fifa_clean['Weight'].isna().sum() > 0:
    print("There are nulls or NA values in the 'Weight' column.")
else:
    print("There are no nulls or NA values in the 'Weight' column.")

    
# This code uses the isnull() and isna() functions to check for nulls and NA values in the "Height" and "Weight" columns of the "fifa_clean" dataframe. If there are any nulls or NA values, it will print a message indicating that there are nulls or NA values in the respective column. Otherwise, it will print a message indicating that there are no nulls or NA values in the respective column.


There are no nulls or NA values in the 'Height' column.
There are no nulls or NA values in the 'Weight' column.


In [42]:
# Let's clean the Preferred Foot column

# Check for nulls or NA values in the "Preferred Foot" column
if fifa_dirty['Preferred Foot'].isnull().sum() > 0 or fifa_dirty['Preferred Foot'].isna().sum() > 0:
    print("There are nulls or NA values in the 'Preferred Foot' column.")
else:
    print("There are no nulls or NA values in the 'Preferred Foot' column.")
    

# Count the number of unique values in the "Preferred Foot" column
num_unique = fifa_dirty['Preferred Foot'].nunique()

# Print the number of unique values
print("Number of unique values in 'Preferred Foot' column:", num_unique)


# Only left and right are the unique values in the Preferred Foot column

There are no nulls or NA values in the 'Preferred Foot' column.
Number of unique values in 'Preferred Foot' column: 2


In [43]:
# Add Preferred Foot column to fifa_clean dataframe

fifa_clean = fifa_clean.join(fifa_dirty['Preferred Foot'])

# Verify that the Preferred Foot column has been added to fifa_clean

print(fifa_clean.info())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 18979 entries, 0 to 18978
Data columns (total 64 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   ID                18979 non-null  int64 
 1   Name              18979 non-null  object
 2   LongName          18979 non-null  object
 3   photoUrl          18979 non-null  object
 4   playerUrl         18979 non-null  object
 5   Nationality       18979 non-null  object
 6   Age               18979 non-null  int64 
 7   OVA               18979 non-null  int64 
 8   POT               18979 non-null  int64 
 9   BOV               18979 non-null  int64 
 10  Attacking         18979 non-null  int64 
 11  Crossing          18979 non-null  int64 
 12  Finishing         18979 non-null  int64 
 13  Heading Accuracy  18979 non-null  int64 
 14  Short Passing     18979 non-null  int64 
 15  Volleys           18979 non-null  int64 
 16  Skill             18979 non-null  int64 
 17  Dribbling   

In [44]:
# Let's clean the Best Position column

# Check for nulls and NAs in the Best Position column
if fifa_dirty['Best Position'].isnull().sum() > 0 or fifa_dirty['Best Position'].isna().sum() > 0:
    print("There are nulls or NA values in the 'Best Position' column.")
else:
    print("There are no nulls or NA values in the 'Best Position' column.")
    

# Count the number of unique values in the "Preferred Foot" column
num_unique = fifa_dirty['Best Position'].nunique()

# Print the number of unique values
print("Number of unique values in 'Best Position' column:", num_unique)


There are no nulls or NA values in the 'Best Position' column.
Number of unique values in 'Best Position' column: 15


In [45]:
# Add Best Position column to fifa_clean dataframe

fifa_clean = fifa_clean.join(fifa_dirty['Best Position'])

# Verify that the Preferred Foot column has been added to fifa_clean

print(fifa_clean.info())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 18979 entries, 0 to 18978
Data columns (total 65 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   ID                18979 non-null  int64 
 1   Name              18979 non-null  object
 2   LongName          18979 non-null  object
 3   photoUrl          18979 non-null  object
 4   playerUrl         18979 non-null  object
 5   Nationality       18979 non-null  object
 6   Age               18979 non-null  int64 
 7   OVA               18979 non-null  int64 
 8   POT               18979 non-null  int64 
 9   BOV               18979 non-null  int64 
 10  Attacking         18979 non-null  int64 
 11  Crossing          18979 non-null  int64 
 12  Finishing         18979 non-null  int64 
 13  Heading Accuracy  18979 non-null  int64 
 14  Short Passing     18979 non-null  int64 
 15  Volleys           18979 non-null  int64 
 16  Skill             18979 non-null  int64 
 17  Dribbling   

In [46]:
# Let's clean the Joined column

# Convert the "Joined" column to datetime datatype
fifa_dirty['Joined'] = pd.to_datetime(fifa_dirty['Joined'])

# Verify the datatype of the "Joined" column has been changed
print(fifa_dirty['Joined'].dtype)


# Check for nulls and NAs in the Joined column
if fifa_dirty['Joined'].isnull().sum() > 0 or fifa_dirty['Joined'].isna().sum() > 0:
    print("There are nulls or NA values in the 'Joined' column.")
else:
    print("There are no nulls or NA values in the 'Joined' column.")
    

datetime64[ns]
There are no nulls or NA values in the 'Joined' column.


In [47]:
# Add Joined column to fifa_clean dataframe

fifa_clean = fifa_clean.join(fifa_dirty['Joined'])

# Verify that the Joined column has been added to fifa_clean

print(fifa_clean.info())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 18979 entries, 0 to 18978
Data columns (total 66 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   ID                18979 non-null  int64         
 1   Name              18979 non-null  object        
 2   LongName          18979 non-null  object        
 3   photoUrl          18979 non-null  object        
 4   playerUrl         18979 non-null  object        
 5   Nationality       18979 non-null  object        
 6   Age               18979 non-null  int64         
 7   OVA               18979 non-null  int64         
 8   POT               18979 non-null  int64         
 9   BOV               18979 non-null  int64         
 10  Attacking         18979 non-null  int64         
 11  Crossing          18979 non-null  int64         
 12  Finishing         18979 non-null  int64         
 13  Heading Accuracy  18979 non-null  int64         
 14  Short Passing     1897

In [48]:
# Let's clean the Loan Date End column

print(fifa_dirty['Loan Date End'])

# We observe that the column has a lot NaN values

# Convert the column datatype to datetime datatype
fifa_dirty['Loan Date End'] = pd.to_datetime(fifa_dirty['Loan Date End'])

print(fifa_dirty['Loan Date End'].dtype)

0        NaN
1        NaN
2        NaN
3        NaN
4        NaN
        ... 
18974    NaN
18975    NaN
18976    NaN
18977    NaN
18978    NaN
Name: Loan Date End, Length: 18979, dtype: object
datetime64[ns]


In [49]:
# Create a new column to indicate players on loan and not on loan. This will justify the NaN values in the Loan Date End column

# Replace the NaN values with 0
fifa_dirty['Loan Date End'].fillna(0, inplace=True)

In [50]:
# Create a new column that indicates if player is on loan or not

fifa_dirty['Loan'] = fifa_dirty['Loan Date End'].apply(lambda x: 'Not on Loan' if x == 0 else 'On Loan')

In [51]:
print(fifa_dirty['Loan'])

0        Not on Loan
1        Not on Loan
2        Not on Loan
3        Not on Loan
4        Not on Loan
            ...     
18974    Not on Loan
18975    Not on Loan
18976    Not on Loan
18977    Not on Loan
18978    Not on Loan
Name: Loan, Length: 18979, dtype: object


In [52]:
# Let's add Loan Date End column to fifa_clean

fifa_clean = fifa_clean.join(fifa_dirty['Loan Date End'])


print(fifa_clean.info())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 18979 entries, 0 to 18978
Data columns (total 67 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   ID                18979 non-null  int64         
 1   Name              18979 non-null  object        
 2   LongName          18979 non-null  object        
 3   photoUrl          18979 non-null  object        
 4   playerUrl         18979 non-null  object        
 5   Nationality       18979 non-null  object        
 6   Age               18979 non-null  int64         
 7   OVA               18979 non-null  int64         
 8   POT               18979 non-null  int64         
 9   BOV               18979 non-null  int64         
 10  Attacking         18979 non-null  int64         
 11  Crossing          18979 non-null  int64         
 12  Finishing         18979 non-null  int64         
 13  Heading Accuracy  18979 non-null  int64         
 14  Short Passing     1897

In [53]:
# Let's add Loan column to fifa_clean

fifa_clean = fifa_clean.join(fifa_dirty['Loan'])


print(fifa_clean.info())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 18979 entries, 0 to 18978
Data columns (total 68 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   ID                18979 non-null  int64         
 1   Name              18979 non-null  object        
 2   LongName          18979 non-null  object        
 3   photoUrl          18979 non-null  object        
 4   playerUrl         18979 non-null  object        
 5   Nationality       18979 non-null  object        
 6   Age               18979 non-null  int64         
 7   OVA               18979 non-null  int64         
 8   POT               18979 non-null  int64         
 9   BOV               18979 non-null  int64         
 10  Attacking         18979 non-null  int64         
 11  Crossing          18979 non-null  int64         
 12  Finishing         18979 non-null  int64         
 13  Heading Accuracy  18979 non-null  int64         
 14  Short Passing     1897

In [54]:
# Let's clean the Value column

print(fifa_dirty['Value'])

# Checkfor null and NA values in the column
if fifa_dirty['Value'].isnull().sum() > 0 or fifa_dirty['Value'].isna().sum() > 0:
    print("There are nulls or NA values in the 'Value' column.")
else:
    print("There are no nulls or NA values in the 'Value' column.")
    
    
    
# Check currency in which the data is in - € 
currencies = fifa_dirty['Value'].str.extract('(\D+)').squeeze().unique()

print(currencies)


0        €103.5M
1           €63M
2          €120M
3          €129M
4          €132M
          ...   
18974      €100K
18975      €130K
18976      €120K
18977      €100K
18978      €100K
Name: Value, Length: 18979, dtype: object
There are no nulls or NA values in the 'Value' column.
['€']


In [55]:
# We observe that some data is in millions and other is in thousands

# Remove currency symbol
fifa_dirty['Value'] = fifa_dirty['Value'].str.replace('€', '')

# Replace K and M by multiplying
fifa_dirty['Value'] = fifa_dirty['Value'].apply(lambda x: float(x[:-1])*1e3 if x[-1]=='K' else (float(x[:-1])*1e6 if x[-1]=='M' else float(x)))

# Handle values with M or K after a decimal point
fifa_dirty['Value'] = fifa_dirty['Value'].apply(lambda x: x*1e6 if 'M' in str(x) and '.' not in str(x) else (x*1e3 if 'K' in str(x) and '.' not in str(x) else x))


In [56]:
print(fifa_dirty['Value'])

0        103500000.0
1         63000000.0
2        120000000.0
3        129000000.0
4        132000000.0
            ...     
18974       100000.0
18975       130000.0
18976       120000.0
18977       100000.0
18978       100000.0
Name: Value, Length: 18979, dtype: float64


In [57]:
# Let's clean the Wage column - We will follow the same steps as in the Value column

print(fifa_dirty['Wage'])

# Checkfor null and NA values in the column
if fifa_dirty['Wage'].isnull().sum() > 0 or fifa_dirty['Wage'].isna().sum() > 0:
    print("There are nulls or NA values in the 'Wage' column.")
else:
    print("There are no nulls or NA values in the 'Wage' column.")
    
    
    
# Check currency in which the data is in - € 
currencies = fifa_dirty['Wage'].str.extract('(\D+)').squeeze().unique()

print(currencies)

0        €560K
1        €220K
2        €125K
3        €370K
4        €270K
         ...  
18974      €1K
18975     €500
18976     €500
18977      €2K
18978      €1K
Name: Wage, Length: 18979, dtype: object
There are no nulls or NA values in the 'Wage' column.
['€']


In [58]:
# We observe that some data is in millions and other is in thousands

# Remove currency symbol
fifa_dirty['Wage'] = fifa_dirty['Wage'].str.replace('€', '')

# Replace K and M by multiplying
fifa_dirty['Wage'] = fifa_dirty['Wage'].apply(lambda x: float(x[:-1])*1e3 if x[-1]=='K' else (float(x[:-1])*1e6 if x[-1]=='M' else float(x)))

# Handle values with M or K after a decimal point
fifa_dirty['Wage'] = fifa_dirty['Wage'].apply(lambda x: x*1e6 if 'M' in str(x) and '.' not in str(x) else (x*1e3 if 'K' in str(x) and '.' not in str(x) else x))


In [59]:
print(fifa_dirty['Wage'])

0        560000.0
1        220000.0
2        125000.0
3        370000.0
4        270000.0
           ...   
18974      1000.0
18975       500.0
18976       500.0
18977      2000.0
18978      1000.0
Name: Wage, Length: 18979, dtype: float64


In [60]:
# Let's clean the Release Clause column - We will follow the same steps as in the Value and Wage column

print(fifa_dirty['Release Clause'])

# Checkfor null and NA values in the column
if fifa_dirty['Release Clause'].isnull().sum() > 0 or fifa_dirty['Release Clause'].isna().sum() > 0:
    print("There are nulls or NA values in the 'Release Clause' column.")
else:
    print("There are no nulls or NA values in the 'Release Clause' column.")
    
    
    
# Check currency in which the data is in - € 
currencies = fifa_dirty['Release Clause'].str.extract('(\D+)').squeeze().unique()

print(currencies)

0        €138.4M
1         €75.9M
2        €159.4M
3          €161M
4        €166.5M
          ...   
18974       €70K
18975      €165K
18976      €131K
18977       €88K
18978       €79K
Name: Release Clause, Length: 18979, dtype: object
There are no nulls or NA values in the 'Release Clause' column.
['€']


In [61]:
# We observe that some data is in millions and other is in thousands

# Remove currency symbol
fifa_dirty['Release Clause'] = fifa_dirty['Release Clause'].str.replace('€', '')

# Replace K and M by multiplying
fifa_dirty['Release Clause'] = fifa_dirty['Release Clause'].apply(lambda x: float(x[:-1])*1e3 if x[-1]=='K' else (float(x[:-1])*1e6 if x[-1]=='M' else float(x)))

# Handle values with M or K after a decimal point
fifa_dirty['Release Clause'] = fifa_dirty['Release Clause'].apply(lambda x: x*1e6 if 'M' in str(x) and '.' not in str(x) else (x*1e3 if 'K' in str(x) and '.' not in str(x) else x))


In [62]:
print(fifa_dirty['Release Clause'])

0        138400000.0
1         75900000.0
2        159400000.0
3        161000000.0
4        166500000.0
            ...     
18974        70000.0
18975       165000.0
18976       131000.0
18977        88000.0
18978        79000.0
Name: Release Clause, Length: 18979, dtype: float64


In [63]:
# Let's add the Value, Wage and Release Clause columns to the fifa_clean dataframe

fifa_clean = fifa_clean.join(fifa_dirty['Value'])


fifa_clean = fifa_clean.join(fifa_dirty['Wage'])


fifa_clean = fifa_clean.join(fifa_dirty['Release Clause'])


In [64]:
print(fifa_clean.info())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 18979 entries, 0 to 18978
Data columns (total 71 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   ID                18979 non-null  int64         
 1   Name              18979 non-null  object        
 2   LongName          18979 non-null  object        
 3   photoUrl          18979 non-null  object        
 4   playerUrl         18979 non-null  object        
 5   Nationality       18979 non-null  object        
 6   Age               18979 non-null  int64         
 7   OVA               18979 non-null  int64         
 8   POT               18979 non-null  int64         
 9   BOV               18979 non-null  int64         
 10  Attacking         18979 non-null  int64         
 11  Crossing          18979 non-null  int64         
 12  Finishing         18979 non-null  int64         
 13  Heading Accuracy  18979 non-null  int64         
 14  Short Passing     1897

In [65]:
# Let's clean the W/F Column

# Check for null and NA values in the column

if fifa_dirty['W/F'].isnull().sum() > 0 or fifa_dirty['W/F'].isna().sum() > 0:
    print("There are nulls or NA values in the 'W/F' column.")
else:
    print("There are no nulls or NA values in the 'W/F' column.")
    
# Check for unique values in the column  
fifa_dirty['W/F'].unique()
    
# We observe that there are no null or NaN values in the W/F column


    

There are no nulls or NA values in the 'W/F' column.


array(['4 ★', '3 ★', '5 ★', '2 ★', '1 ★'], dtype=object)

In [66]:
# Let us remove the stars in the column
# Since the column represents Weak Foot rating on a scale of 1-5, we can use the str.replace() function to replace the star with an empty string

fifa_dirty['W/F'] = fifa_dirty['W/F'].str.replace('★','')


In [67]:
# Remove the space left after removing the star
fifa_dirty['W/F'] = fifa_dirty['W/F'].str.replace('★', '').str.strip()

fifa_dirty['W/F'].unique()


array(['4', '3', '5', '2', '1'], dtype=object)

In [68]:
# Let's clean the SM Column

# Check for null and NA values in the column

if fifa_dirty['SM'].isnull().sum() > 0 or fifa_dirty['SM'].isna().sum() > 0:
    print("There are nulls or NA values in the 'SM' column.")
else:
    print("There are no nulls or NA values in the 'SM' column.")
    
# Check for unique values in the column  
fifa_dirty['SM'].unique()
    
# We observe that there are no null or NaN values in the SM column


There are no nulls or NA values in the 'SM' column.


array(['4★', '5★', '1★', '2★', '3★'], dtype=object)

In [69]:
# Let us remove the stars in the column
# Since the column represents SM rating on a scale of 1-5, we can use the str.replace() function to replace the star with an empty string

fifa_dirty['SM'] = fifa_dirty['SM'].str.replace('★','')

fifa_dirty['W/F'] = fifa_dirty['W/F'].str.replace('★', '').str.strip()

fifa_dirty['W/F'].unique()


array(['4', '3', '5', '2', '1'], dtype=object)

In [70]:
# Let's clean the SM Column

# Check for null and NA values in the column

if fifa_dirty['A/W'].isnull().sum() > 0 or fifa_dirty['A/W'].isna().sum() > 0:
    print("There are nulls or NA values in the 'A/W' column.")
else:
    print("There are no nulls or NA values in the 'A/W' column.")
    
# Check for unique values in the column  
fifa_dirty['A/W'].unique()
    
# We observe that there are no null or NaN values in the A/W column


There are no nulls or NA values in the 'A/W' column.


array(['Medium', 'High', 'Low'], dtype=object)

In [71]:
# Let's clean the D/W Column

# Check for null and NA values in the column

if fifa_dirty['D/W'].isnull().sum() > 0 or fifa_dirty['D/W'].isna().sum() > 0:
    print("There are nulls or NA values in the 'D/W' column.")
else:
    print("There are no nulls or NA values in the 'D/W' column.")
    
# Check for unique values in the column  
fifa_dirty['D/W'].unique()
    
# We observe that there are no null or NaN values in the A/W column


There are no nulls or NA values in the 'D/W' column.


array(['Low', 'Medium', 'High'], dtype=object)

In [72]:
# Let's clean the IR Column

# Check for null and NA values in the column

if fifa_dirty['IR'].isnull().sum() > 0 or fifa_dirty['IR'].isna().sum() > 0:
    print("There are nulls or NA values in the 'IR' column.")
else:
    print("There are no nulls or NA values in the 'IR' column.")
    
# Check for unique values in the column  
fifa_dirty['IR'].unique()
    
# We observe that there are no null or NaN values in the A/W column


There are no nulls or NA values in the 'IR' column.


array(['5 ★', '3 ★', '4 ★', '2 ★', '1 ★'], dtype=object)

In [73]:
# Let us remove the stars in the column
# Since the column represents IR rating on a scale of 1-5, we can use the str.replace() function to replace the star with an empty string

fifa_dirty['IR'] = fifa_dirty['IR'].str.replace('★','')

fifa_dirty['IR'] = fifa_dirty['IR'].str.replace('★', '').str.strip()

fifa_dirty['IR'].unique()

array(['5', '3', '4', '2', '1'], dtype=object)

In [74]:
# Let's add the W/F, SM, A/W, D/W, IR columns to the fifa_clean dataframe

fifa_clean = fifa_clean.join(fifa_dirty['W/F'])


fifa_clean = fifa_clean.join(fifa_dirty['SM'])


fifa_clean = fifa_clean.join(fifa_dirty['A/W'])

fifa_clean = fifa_clean.join(fifa_dirty['D/W'])

fifa_clean = fifa_clean.join(fifa_dirty['IR'])


In [75]:
print(fifa_clean.info())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 18979 entries, 0 to 18978
Data columns (total 76 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   ID                18979 non-null  int64         
 1   Name              18979 non-null  object        
 2   LongName          18979 non-null  object        
 3   photoUrl          18979 non-null  object        
 4   playerUrl         18979 non-null  object        
 5   Nationality       18979 non-null  object        
 6   Age               18979 non-null  int64         
 7   OVA               18979 non-null  int64         
 8   POT               18979 non-null  int64         
 9   BOV               18979 non-null  int64         
 10  Attacking         18979 non-null  int64         
 11  Crossing          18979 non-null  int64         
 12  Finishing         18979 non-null  int64         
 13  Heading Accuracy  18979 non-null  int64         
 14  Short Passing     1897

In [77]:
# Let's clean the Hits Column

# Check for null and NA values in the column

if fifa_dirty['Hits'].isnull().sum() > 0 or fifa_dirty['Hits'].isna().sum() > 0:
    print("There are nulls or NA values in the 'Hits' column.")
else:
    print("There are no nulls or NA values in the 'Hits' column.")
    
# Check for unique values in the column  
fifa_dirty['Hits'].unique()
    
# We observe that there are nulls or NaN values in the Hits column and the data here is a bit messy
# We also notice that some data in the column is in thousands in the form of K

# Let's convert the K in thousands to 000s
# replace the 'K' suffix with '000' in the "Hits" column
fifa_dirty["Hits"] = fifa_dirty["Hits"].str.replace('K', '000')

# convert the "Hits" column to numeric type
fifa_dirty["Hits"] = pd.to_numeric(fifa_dirty["Hits"], errors='coerce')

# print the first 10 rows of the modified "Hits" column
print(fifa_dirty["Hits"].head(10))


There are nulls or NA values in the 'Hits' column.
0    771.0
1    562.0
2    150.0
3    207.0
4    595.0
5    248.0
6    246.0
7    120.0
8      1.6
9    130.0
Name: Hits, dtype: float64


In [78]:
# count the number of null and NaN values in the "Hits" column
num_null = fifa_dirty["Hits"].isnull().sum()
num_nan = np.isnan(fifa_dirty["Hits"]).sum()

# print the results
print("Number of null values:", num_null)
print("Number of NaN values:", num_nan)


Number of null values: 2595
Number of NaN values: 2595


In [79]:
# We can replace the Null values with 0 in the Hits column

fifa_dirty["Hits"] = fifa_dirty["Hits"].fillna(0)


# print the results
print(fifa_dirty['Hits'])


0        771.0
1        562.0
2        150.0
3        207.0
4        595.0
         ...  
18974      0.0
18975      0.0
18976      0.0
18977      0.0
18978      0.0
Name: Hits, Length: 18979, dtype: float64


In [80]:
# Let's add the Hits column to the fifa_clean dataframe

fifa_clean = fifa_clean.join(fifa_dirty['Hits'])


print(fifa_clean.info())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 18979 entries, 0 to 18978
Data columns (total 77 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   ID                18979 non-null  int64         
 1   Name              18979 non-null  object        
 2   LongName          18979 non-null  object        
 3   photoUrl          18979 non-null  object        
 4   playerUrl         18979 non-null  object        
 5   Nationality       18979 non-null  object        
 6   Age               18979 non-null  int64         
 7   OVA               18979 non-null  int64         
 8   POT               18979 non-null  int64         
 9   BOV               18979 non-null  int64         
 10  Attacking         18979 non-null  int64         
 11  Crossing          18979 non-null  int64         
 12  Finishing         18979 non-null  int64         
 13  Heading Accuracy  18979 non-null  int64         
 14  Short Passing     1897

In [81]:
# Let's clean the Contract column

# Check for null and NA values in the column

if fifa_dirty['Contract'].isnull().sum() > 0 or fifa_dirty['Contract'].isna().sum() > 0:
    print("There are nulls or NA values in the 'Contract' column.")
else:
    print("There are no nulls or NA values in the 'Contract' column.")
    
# Check for unique values in the column  
fifa_dirty['Contract'].unique()


There are no nulls or NA values in the 'Contract' column.


array(['2004 ~ 2021', '2018 ~ 2022', '2014 ~ 2023', '2015 ~ 2023',
       '2017 ~ 2022', '2017 ~ 2023', '2018 ~ 2024', '2014 ~ 2022',
       '2018 ~ 2023', '2016 ~ 2023', '2013 ~ 2023', '2011 ~ 2023',
       '2009 ~ 2022', '2005 ~ 2021', '2011 ~ 2021', '2015 ~ 2022',
       '2017 ~ 2024', '2010 ~ 2024', '2012 ~ 2021', '2019 ~ 2024',
       '2015 ~ 2024', '2017 ~ 2025', '2020 ~ 2025', '2019 ~ 2023',
       '2008 ~ 2023', '2015 ~ 2021', '2020 ~ 2022', '2012 ~ 2022',
       '2016 ~ 2025', '2013 ~ 2022', '2011 ~ 2022', '2012 ~ 2024',
       '2016 ~ 2021', '2012 ~ 2023', '2008 ~ 2022', '2019 ~ 2022',
       '2017 ~ 2021', '2013 ~ 2024', '2020 ~ 2024', '2010 ~ 2022',
       '2020 ~ 2021', '2011 ~ 2024', '2020 ~ 2023', '2014 ~ 2024',
       '2013 ~ 2026', '2016 ~ 2022', '2010 ~ 2021', '2013 ~ 2021',
       '2019 ~ 2025', '2018 ~ 2025', '2016 ~ 2024', '2018 ~ 2021',
       '2009 ~ 2024', '2007 ~ 2022', 'Jun 30, 2021 On Loan',
       '2009 ~ 2021', '2019 ~ 2021', '2019 ~ 2026', 'Free', '2012 ~ 

In [82]:
#explore the players with free contract

fifa_dirty[fifa_dirty.Contract=='Free']


Age  OVA  POT     Club Contract  Positions Height Weight  \
289     20   81   81  No Club     Free      LB/LM  178cm   69kg   
292     24   81   81  No Club     Free     CB/CDM  181cm   82kg   
369     36   80   80  No Club     Free     CB/CDM  181cm   82kg   
374     32   80   80  No Club     Free         GK  190cm   85kg   
375     32   80   80  No Club     Free      LB/LM  178cm   69kg   
...    ...  ...  ...      ...      ...        ...    ...    ...   
17220   22   56   64  No Club     Free         RB  172cm   67kg   
17343   20   56   56  No Club     Free      RM/CM  168cm   62kg   
17659   27   55   55  No Club     Free         ST  178cm   77kg   
17661   28   55   55  No Club     Free  RM/CAM/RB  175cm   73kg   
18205   26   53   54  No Club     Free         LM  171cm   65kg   

      Preferred Foot  BOV  ...     D/W IR PAC  SHO  PAS  DRI  DEF  PHY   Hits  \
289             Left   81  ...  Medium  1  79   54   76   78   75   74  172.0   
292            Right   81  ...  Medium  1  71   61   57   62   82   80   75.0   
369            Right   80  ...  Medium  1  71   61   57   62   82   75   11.0   
374            Right   80  ...  Medium  1  78   81   77   77   42   80   18.0   
375             Left   80  ...  Medium  1  79   54   75   78   76   74   17.0   
...              ...  ...  ...     ... ..  ..  ...  ...  ...  ...  ...    ...   
17220          Right   56  ...  Medium  1  73   33   45   55   51   55    0.0   
17343          Right   56  ...     Low  1  79   42   50   59   29   49    0.0   
17659          Right   57  ...  Medium  1  67   52   47   61   32   57    0.0   
17661           Left   55  ...  Medium  1  73   49   51   54   41   70    0.0   
18205           Left   53  ...  Medium  1  68   41   45   58   22   41    0.0   

              Loan  
289    Not on Loan  
292    Not on Loan  
369    Not on Loan  
374    Not on Loan  
375    Not on Loan  
...            ...  
17220  Not on Loan  
17343  Not on Loan  
17659  Not on Loan  
17661  Not on Loan  
18205  Not on Loan  

[237 rows x 72 columns]

In [84]:
# Explore players that are on loan

on_loan=fifa_dirty[fifa_dirty.Contract.str.contains("On Loan")]
on_loan

Age  OVA  POT                           Club              Contract  \
205     30   83   83              Tottenham Hotspur  Jun 30, 2021 On Loan   
248     27   82   86                         Fulham  Jun 30, 2021 On Loan   
254     28   82   82            Paris Saint-Germain  Jun 30, 2021 On Loan   
302     26   81   81                         Napoli  Jun 30, 2021 On Loan   
306     24   81   85                Atlético Madrid  Jun 30, 2021 On Loan   
...    ...  ...  ...                            ...                   ...   
18472   20   52   65                   Macarthur FC  Aug 31, 2021 On Loan   
18571   20   51   63                      Stevenage  Jun 30, 2021 On Loan   
18600   23   51   60  Shijiazhuang Ever Bright F.C.  Dec 31, 2020 On Loan   
18622   23   51   60  Shijiazhuang Ever Bright F.C.  Dec 31, 2020 On Loan   
18680   25   51   56               Guangzhou R&F FC  Dec 31, 2020 On Loan   

      Positions Height Weight Preferred Foot  BOV  ...     D/W IR PAC  SHO  \
205          RW  185cm   82kg           Left   83  ...  Medium  4  88   84   
248          GK  195cm   94kg          Right   82  ...  Medium  2  85   79   
254      CDM/CM  188cm   83kg          Right   82  ...    High  3  66   63   
302    RM/RW/ST  171cm   67kg           Left   81  ...  Medium  2  83   77   
306      CDM/CM  166cm   60kg          Right   82  ...    High  2  74   67   
...         ...    ...    ...            ...  ...  ...     ... ..  ..  ...   
18472        ST  182cm   72kg           Left   54  ...     Low  1  67   51   
18571  CM/RW/LW  178cm   64kg           Left   55  ...  Medium  1  57   45   
18600        CM  181cm   65kg          Right   55  ...  Medium  1  71   42   
18622    CAM/LM  177cm   68kg          Right   53  ...  Medium  1  59   41   
18680  CM/LM/LB  177cm   70kg           Left   59  ...  Medium  1  61   35   

       PAS  DRI  DEF  PHY   Hits     Loan  
205     82   80   58   73  282.0  On Loan  
248     80   84   55   78   94.0  On Loan  
254     72   73   80   85  285.0  On Loan  
302     78   84   42   55   93.0  On Loan  
306     75   79   79   74  194.0  On Loan  
...    ...  ...  ...  ...    ...      ...  
18472   39   53   21   52    0.0  On Loan  
18571   52   59   38   48    0.0  On Loan  
18600   52   54   47   52    0.0  On Loan  
18622   55   49   34   47    0.0  On Loan  
18680   49   50   59   58    0.0  On Loan  

[1013 rows x 72 columns]

In [94]:
# From the exploration, it shows their contract is the same as loan end date which is kind of irrelevant keeping both

# Check the number of players that are on loan
print(len(on_loan))

# Count the number of non-zero values in 'Loan Date End' column
num_loans = (fifa_dirty['Loan Date End'] != 0).astype(bool).sum()

# Print the result
print(f"There are {num_loans} non-zero values in the 'Loan Date End' column.")


1013
There are 1013 non-zero values in the 'Loan Date End' column.


In [95]:
# Define the contract_sorting() function

def contract_sorting(Contract):
    if Contract == 'Free':
        contract_type = 'Free Agent'
        contract_start = None
        contract_end = None
    elif ' On Loan' in Contract:
        contract_type = 'On Loan'
        contract_start= None
        contract_end= Contract.replace(' On Loan','')
    else:
        contract_type = 'On Contract'
        contract_start, contract_end = Contract.split(' ~ ')

    return contract_start, contract_end,contract_type

# Apply the function to the 'Contract' column of the fifa_dirty dataframe
fifa_dirty[['Contract Start', 'Contract End', 'Contract Type']] = fifa_dirty['Contract'].apply(lambda x: pd.Series(contract_sorting(x)))

# Print the first few rows of the modified dataframe to check the result
print(fifa_dirty.head())


   Age  OVA  POT                 Club     Contract Positions Height Weight  \
0   33   93   93         FC Barcelona  2004 ~ 2021  RW/ST/CF  170cm   72kg   
1   35   92   92             Juventus  2018 ~ 2022     ST/LW  187cm   83kg   
2   27   91   93      Atlético Madrid  2014 ~ 2023        GK  188cm   87kg   
3   29   91   91      Manchester City  2015 ~ 2023    CAM/CM  181cm   70kg   
4   28   91   91  Paris Saint-Germain  2017 ~ 2022    LW/CAM  175cm   68kg   

  Preferred Foot  BOV  ... SHO PAS DRI  DEF  PHY   Hits         Loan  \
0           Left   93  ...  92  91  95   38   65  771.0  Not on Loan   
1          Right   92  ...  93  81  89   35   77  562.0  Not on Loan   
2          Right   91  ...  92  78  90   52   90  150.0  Not on Loan   
3          Right   91  ...  86  93  88   64   78  207.0  Not on Loan   
4          Right   91  ...  85  86  94   36   59  595.0  Not on Loan   

   Contract Start  Contract End  Contract Type  
0            2004          2021    On Contract  


In [99]:
print(fifa_dirty['Contract Start'])

print(fifa_dirty['Contract End'])

print(fifa_dirty['Contract Type'])

0        2004
1        2018
2        2014
3        2015
4        2017
         ... 
18974    2018
18975    2020
18976    2019
18977    2020
18978    2019
Name: Contract Start, Length: 18979, dtype: object
0        2021
1        2022
2        2023
3        2023
4        2022
         ... 
18974    2022
18975    2021
18976    2020
18977    2022
18978    2023
Name: Contract End, Length: 18979, dtype: object
0        On Contract
1        On Contract
2        On Contract
3        On Contract
4        On Contract
            ...     
18974    On Contract
18975    On Contract
18976    On Contract
18977    On Contract
18978    On Contract
Name: Contract Type, Length: 18979, dtype: object


In [101]:
# Using the apply method on contract_sorting and apply method to put it into three series and assigning the values to newly created columns

fifa_dirty[['Contract Start','Contract End', 'Contract Type']]=fifa_dirty.Contract.apply(contract_sorting).apply(pd.Series)

In [103]:
fifa_clean = fifa_clean.join(fifa_dirty['Contract Start'])

fifa_clean = fifa_clean.join(fifa_dirty['Contract End'])

fifa_clean = fifa_clean.join(fifa_dirty['Contract Type'])


print(fifa_clean.info())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 18979 entries, 0 to 18978
Data columns (total 80 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   ID                18979 non-null  int64         
 1   Name              18979 non-null  object        
 2   LongName          18979 non-null  object        
 3   photoUrl          18979 non-null  object        
 4   playerUrl         18979 non-null  object        
 5   Nationality       18979 non-null  object        
 6   Age               18979 non-null  int64         
 7   OVA               18979 non-null  int64         
 8   POT               18979 non-null  int64         
 9   BOV               18979 non-null  int64         
 10  Attacking         18979 non-null  int64         
 11  Crossing          18979 non-null  int64         
 12  Finishing         18979 non-null  int64         
 13  Heading Accuracy  18979 non-null  int64         
 14  Short Passing     1897

In [106]:
# Dropping irrelevant columns from fifa_clean

fifa_clean.drop(['Loan Date End', 'Loan'],axis=1,inplace=True)


In [108]:
print(fifa_clean.info())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 18979 entries, 0 to 18978
Data columns (total 78 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   ID                18979 non-null  int64         
 1   Name              18979 non-null  object        
 2   LongName          18979 non-null  object        
 3   photoUrl          18979 non-null  object        
 4   playerUrl         18979 non-null  object        
 5   Nationality       18979 non-null  object        
 6   Age               18979 non-null  int64         
 7   OVA               18979 non-null  int64         
 8   POT               18979 non-null  int64         
 9   BOV               18979 non-null  int64         
 10  Attacking         18979 non-null  int64         
 11  Crossing          18979 non-null  int64         
 12  Finishing         18979 non-null  int64         
 13  Heading Accuracy  18979 non-null  int64         
 14  Short Passing     1897

We have managed to clean all the relevant columns in the dataset and get it ready for exploration, analysis and visualization.

In [121]:
# Rename some columns 
fifa_clean.rename(columns={'Height':'Height (cm)',
                  'Weight':'Weight (kg)',
                  'Value':'Value (Euro)',
                  'Wage':'Wage (Euro)',
                  'Release Clause':'Release Clause (Euro)'}, inplace=True)

In [122]:
# Save the fifa_clean dataframe as a CSV file to the specified path
fifa_clean.to_csv('C:\\Users\\ragir\\OneDrive\\Desktop\\DATA ANALYTICS\\Portfolio Projects\\Data Cleaning Challenge\\fifa_clean.csv', index=False)

# Print a message to confirm the file has been saved
print('fifa_clean.csv saved successfully.')


fifa_clean.csv saved successfully.


We have exported the clean data as a csv file which we can now analyze in any other tool easily and visualize the data in BI tools as well.